
 # AI-Powered Retail Intelligence Platform
- Company: NovaMart
-  Dataset: Olist Brazilian E-Commerce Dataset


In [2]:
print("="*70)
print(" AI-Powered Retail Intelligence Platform")
print(" Company: NovaMart")
print(" Dataset: Olist Brazilian E-Commerce Dataset")
print("="*70)

 AI-Powered Retail Intelligence Platform
 Company: NovaMart
 Dataset: Olist Brazilian E-Commerce Dataset


# Phase 1:	Environment Setup & Data Acquisition	

## Step 1: Import Libraries

In [1]:
# Data Manipulation
import pandas as pd
import numpy as np

# Visualization
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

# Machine Learning
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

# Ignore warnings
import warnings
warnings.filterwarnings('ignore')

# Display Settings
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 100)

print("Libraries Imported Successfully!")

Libraries Imported Successfully!


## Download Dataset

In [7]:
import os

DATA_PATH = r"C:\Users\tiwal\Desktop\OSAP\CIMT\CAPSTONE II\archive"

## Step 4: Load All Datasets

In [8]:
customers = pd.read_csv(os.path.join(DATA_PATH,
                                     "olist_customers_dataset.csv"))

geolocation = pd.read_csv(os.path.join(DATA_PATH,
                                       "olist_geolocation_dataset.csv"))

order_items = pd.read_csv(os.path.join(DATA_PATH,
                                       "olist_order_items_dataset.csv"))

payments = pd.read_csv(os.path.join(DATA_PATH,
                                    "olist_order_payments_dataset.csv"))

reviews = pd.read_csv(os.path.join(DATA_PATH,
                                   "olist_order_reviews_dataset.csv"))

orders = pd.read_csv(os.path.join(DATA_PATH,
                                  "olist_orders_dataset.csv"))

products = pd.read_csv(os.path.join(DATA_PATH,
                                    "olist_products_dataset.csv"))

sellers = pd.read_csv(os.path.join(DATA_PATH,
                                   "olist_sellers_dataset.csv"))

translation = pd.read_csv(os.path.join(DATA_PATH,
                         "product_category_name_translation.csv"))

## Step 5: Verify All Datasets Loaded

In [9]:
datasets = {
    "Customers": customers,
    "Geolocation": geolocation,
    "Order Items": order_items,
    "Payments": payments,
    "Reviews": reviews,
    "Orders": orders,
    "Products": products,
    "Sellers": sellers,
    "Category Translation": translation
}

for name, df in datasets.items():
    print(f"{name:25} {df.shape}")

Customers                 (99441, 5)
Geolocation               (1000163, 5)
Order Items               (112650, 7)
Payments                  (103886, 5)
Reviews                   (99224, 7)
Orders                    (99441, 8)
Products                  (32951, 9)
Sellers                   (3095, 4)
Category Translation      (71, 2)


## Step 6: Display First Five Rows

In [10]:
for name, df in datasets.items():

    print("="*80)
    print(name)
    print("="*80)

    display(df.head())

Customers


,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


Geolocation


,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP
2,1046,-23.546129,-46.642951,sao paulo,SP
3,1041,-23.544392,-46.639499,sao paulo,SP
4,1035,-23.541578,-46.641607,sao paulo,SP


Order Items


,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14


Payments


,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71
3,ba78997921bbcdc1373bb41e913ab953,1,credit_card,8,107.78
4,42fdf880ba16b47b59251dd489d4441a,1,credit_card,2,128.45


Reviews


,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,2018-02-17 00:00:00,2018-02-18 14:36:24
3,e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,NaN,Recebi bem antes do prazo estipulado.,2017-04-21 00:00:00,2017-04-21 22:02:06
4,f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,NaN,Parabéns lojas lannister adorei comprar pela I...,2018-03-01 00:00:00,2018-03-02 10:26:53


Orders


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


Products


,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.0,261.0,1.0,371.0,26.0,4.0,26.0
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.0,402.0,4.0,625.0,20.0,17.0,13.0


Sellers


,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ
3,c0f3eea2e14555b6faeea3dd58c1b1c3,4195,sao paulo,SP
4,51a04a8a6bdcb23deccc82b0b80742cf,12914,braganca paulista,SP


Category Translation


,product_category_name,product_category_name_english
0,beleza_saude,health_beauty
1,informatica_acessorios,computers_accessories
2,automotivo,auto
3,cama_mesa_banho,bed_bath_table
4,moveis_decoracao,furniture_decor


## Step 7: Dataset Summary

In [11]:
summary = pd.DataFrame()
summary["Dataset"] = datasets.keys()
summary["Rows"] = [df.shape[0] for df in datasets.values()]
summary["Columns"] = [df.shape[1] for df in datasets.values()]

summary

,Dataset,Rows,Columns
0,Customers,99441,5
1,Geolocation,1000163,5
2,Order Items,112650,7
3,Payments,103886,5
4,Reviews,99224,7
5,Orders,99441,8
6,Products,32951,9
7,Sellers,3095,4
8,Category Translation,71,2


## Step 8: Dataset Information

In [12]:
for name, df in datasets.items():

    print("="*80)
    print(name)
    print("="*80)

    print(df.info())

Customers
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB
None
Geolocation
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  int64  
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng              1000163 non-null  float64
 3   geolocation_city    

## Step 9: Check Missing Values

In [13]:
missing = pd.DataFrame()

missing["Dataset"] = datasets.keys()

missing["Missing Values"] = [
    df.isnull().sum().sum()
    for df in datasets.values()
]

missing

,Dataset,Missing Values
0,Customers,0
1,Geolocation,0
2,Order Items,0
3,Payments,0
4,Reviews,145903
5,Orders,4908
6,Products,2448
7,Sellers,0
8,Category Translation,0


## To find out which columns in Review, Orders and Products contain missing values

In [14]:
for name, df in datasets.items():

    print("="*70)
    print(name)
    print("="*70)

    print(df.isnull().sum())

    print("\n")

Customers
customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64


Geolocation
geolocation_zip_code_prefix    0
geolocation_lat                0
geolocation_lng                0
geolocation_city               0
geolocation_state              0
dtype: int64


Order Items
order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64


Payments
order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64


Reviews
review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_comment_message     58247
review_creation_date           0
review_answer_timestamp        0
dtype: int64


Orders
order_id        

In [15]:
orders["order_status"].value_counts()

delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: order_status, dtype: int64

In [16]:
orders["order_status"].unique()

array(['delivered', 'invoiced', 'shipped', 'processing', 'unavailable',
       'canceled', 'created', 'approved'], dtype=object)

In [ ]:
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2

In [18]:
products["product_category_name"].nunique()

73

# Phase 2: Data Cleaning 

### Reviews

In [19]:
reviews["review_comment_title"] = (
    reviews["review_comment_title"]
    .fillna("")
)

reviews["review_comment_message"] = (
    reviews["review_comment_message"]
    .fillna("")
)

### Products

In [20]:
products["product_category_name"] = (
    products["product_category_name"]
    .fillna("Unknown")
)

numeric_columns = [

    "product_name_lenght",
    "product_description_lenght",
    "product_photos_qty",
    "product_weight_g",
    "product_length_cm",
    "product_height_cm",
    "product_width_cm"

]

for col in numeric_columns:

    products[col].fillna(
        products[col].median(),
        inplace=True
    )

### Orders

### For now, I will leave the missing timestamps unchanged. Before deciding how to handle them, examine whether they are associated with particular order statuses.

In [21]:
orders.groupby("order_status")[
    [
        "order_approved_at",
        "order_delivered_carrier_date",
        "order_delivered_customer_date"
    ]
].apply(lambda x: x.isnull().sum())

,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date
order_status,,,
approved,0,2,2
canceled,141,550,619
created,5,5,5
delivered,14,2,8
invoiced,0,314,314
processing,0,301,301
shipped,0,0,1107
unavailable,0,609,609


* The missing timestamp values in the orders dataset were examined by order_status. The analysis showed that the majority of missing values occurred in orders that were canceled, unavailable, processing, invoiced, or newly created. These missing timestamps reflect the normal order lifecycle rather than data quality issues and were therefore retained. A very small number of delivered orders (<0.02%) contained missing timestamps, likely due to data-entry inconsistencies. These records were excluded only from delivery-time analyses while being retained for other analyses.*

| Order Status | Missing `order_approved_at` | Missing `order_delivered_carrier_date` | Missing `order_delivered_customer_date` | Interpretation                                                             |
| ------------ | --------------------------: | -------------------------------------: | --------------------------------------: | -------------------------------------------------------------------------- |
| approved     |                           0 |                                      2 |                                       2 | Recently approved; not yet shipped or delivered.                           |
| canceled     |                         141 |                                    550 |                                     619 | Expected. Many canceled orders were never approved, shipped, or delivered. |
| created      |                           5 |                                      5 |                                       5 | Order was created but never progressed.                                    |
| delivered    |                          14 |                                      2 |                                       8 | Very small number of anomalies that should be investigated.                |
| invoiced     |                           0 |                                    314 |                                     314 | Invoice generated; shipment has not started.                               |
| processing   |                           0 |                                    301 |                                     301 | Still being processed.                                                     |
| shipped      |                           0 |                                      0 |                                    1107 | Already shipped but not yet delivered.                                     |
| unavailable  |                           0 |                                    609 |                                     609 | Product unavailable, so shipment and delivery never occurred.              |


#### 1. Most missing values are legitimate

The majority of missing timestamps occur in:

- created
- processing
- invoiced
- canceled
- unavailable

These orders never reached the shipment or delivery stages, so missing timestamps are expected.

#### 2. The "delivered" orders deserve attention

There are:

- 14 delivered orders with a missing approval timestamp.
- 2 delivered orders with a missing carrier timestamp.
- 8 delivered orders with a missing customer delivery timestamp.

Since these are out of 96,478 delivered orders, they represent a tiny fraction:

- Missing approval timestamp: about 0.015%
- Missing carrier timestamp: about 0.002%
- Missing delivery timestamp: about 0.008%

These are most likely data-entry issues or inconsistencies.

#### Action Taken
#### order_approved_at

- Keep the missing values for non-completed orders.

For the 14 delivered orders, there two reasonable options:

- Drop those 14 records if I need complete timestamp information.
- Retain them if approval time is not required for a particular model.

#### order_delivered_carrier_date

Keep missing values for:

canceled
created
invoiced
processing
unavailable

Investigate the 2 delivered orders separately. I can either exclude them from delivery-time calculations or leave them out of delivery models.

#### order_delivered_customer_date

Keep missing values for all non-delivered statuses.

For the 8 delivered orders with missing delivery dates, exclude them when computing delivery duration because the duration cannot be calculated reliably.

### Step 10: Check Duplicate Records

In [22]:
duplicates = pd.DataFrame()

duplicates["Dataset"] = datasets.keys()

duplicates["Duplicate Rows"] = [
    df.duplicated().sum()
    for df in datasets.values()
]

duplicates

,Dataset,Duplicate Rows
0,Customers,0
1,Geolocation,261831
2,Order Items,0
3,Payments,0
4,Reviews,0
5,Orders,0
6,Products,0
7,Sellers,0
8,Category Translation,0


In [23]:
# Step 1: Check exact duplicate rows
geolocation.duplicated().sum()

261831

### These are exact duplicates (every column is identical).

In [26]:
# Remove exact duplicates
geolocation = geolocation.drop_duplicates()

print("New Shape:", geolocation.shape)

New Shape: (738332, 5)


In [27]:
# Verify
geolocation.duplicated().sum()

0

In [39]:
geolocation.head()

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP
2,1046,-23.546129,-46.642951,sao paulo,SP
3,1041,-23.544392,-46.639499,sao paulo,SP
4,1035,-23.541578,-46.641607,sao paulo,SP


In [33]:
# Check for duplicate ZIP code prefixes

# This is more important.

geolocation["geolocation_zip_code_prefix"].duplicated().sum()

719317

#### These are not duplicates—they represent different locations within the same ZIP code prefix.

### Preprocessing

If using geolocation to map customers or sellers, create a single representative coordinate for each ZIP code prefix.


In [40]:
#Number of cities per ZIP
city_counts = (
    geolocation.groupby("geolocation_zip_code_prefix")
    ["geolocation_city"]
    .nunique()
)

city_counts.value_counts().sort_index()

1    10459
2     8265
3      255
4       27
5        9
Name: geolocation_city, dtype: int64

In [38]:
#Number of states per ZIP
state_counts = (
    geolocation
    .groupby("geolocation_zip_code_prefix")["geolocation_state"]
    .nunique()
)

state_counts.value_counts().sort_index()

1    19007
2        8
Name: geolocation_state, dtype: int64

In [41]:
#Final aggregation 
geo_clean = (
    geolocation
    .groupby("geolocation_zip_code_prefix", as_index=False)
    .agg({
        "geolocation_lat": "mean",
        "geolocation_lng": "mean",
        "geolocation_city": lambda x: x.mode().iloc[0],
        "geolocation_state": lambda x: x.mode().iloc[0]
    })
)

In [42]:
geo_clean.head()

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1001,-23.550227,-46.634039,sao paulo,SP
1,1002,-23.547657,-46.634991,sao paulo,SP
2,1003,-23.549000,-46.635582,sao paulo,SP
3,1004,-23.549829,-46.634792,sao paulo,SP
4,1005,-23.549547,-46.636406,sao paulo,SP


The geolocation dataset contained multiple records for the same ZIP code prefix because several addresses may share a common prefix. Exact duplicate records were removed. The remaining records were aggregated to one representative location per ZIP code prefix by averaging latitude and longitude, while the most frequently occurring city and state (mode) were retained. An exploratory analysis showed that 99.96% of ZIP code prefixes mapped to a single state, indicating that the aggregation preserved the geographic integrity of the dataset.

In [49]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       99441 non-null  object        
 1   customer_id                    99441 non-null  object        
 2   order_status                   99441 non-null  object        
 3   order_purchase_timestamp       99441 non-null  datetime64[ns]
 4   order_approved_at              99281 non-null  datetime64[ns]
 5   order_delivered_carrier_date   97658 non-null  datetime64[ns]
 6   order_delivered_customer_date  96476 non-null  datetime64[ns]
 7   order_estimated_delivery_date  99441 non-null  datetime64[ns]
dtypes: datetime64[ns](5), object(3)
memory usage: 6.1+ MB


## Verify Cleaning

In [52]:
for name,df in datasets.items():

    print("="*70)

    print(name)

    print("="*70)

    print(df.isnull().sum())

    print()

Customers
customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

Geolocation
geolocation_zip_code_prefix    0
geolocation_lat                0
geolocation_lng                0
geolocation_city               0
geolocation_state              0
dtype: int64

Order Items
order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

Payments
order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

Reviews
review_id                  0
order_id                   0
review_score               0
review_comment_title       0
review_comment_message     0
review_creation_date       0
review_answer_timestamp    0
dtype: int64

Orders
order_id                            0
customer_id

# Save Cleaned Data

In [55]:
customers.to_csv("customers_clean.csv",index=False)

orders.to_csv("orders_clean.csv",index=False)

products.to_csv("products_clean.csv",index=False)

reviews.to_csv("reviews_clean.csv",index=False)

geo_clean.to_csv("geolocation_clean.csv",index=False)

payments.to_csv("payments_clean.csv",index=False)

order_items.to_csv("order_items_clean.csv",index=False)

sellers.to_csv("sellers_clean.csv",index=False)

translation.to_csv("category_translation_clean.csv",index=False)

# Data Integration

## Dataset Relationship Diagram

                           CUSTOMERS
                               │
                      customer_id
                               │
                               ▼
                            ORDERS
                               │
          ┌──────────────┬─────┴─────────────┐
          │              │                   │
          ▼              ▼                   ▼
      PAYMENTS       REVIEWS          ORDER_ITEMS
                                               │
                                    ┌──────────┴──────────┐
                                    ▼                     ▼
                               PRODUCTS              SELLERS
                                    │                     │
                                    ▼                     ▼
                         CATEGORY_TRANSLATION     GEOLOCATION
                                    ▲
                                    │
                           (Customers also link
                           through ZIP code)

## Objective of Data Integration

The goal is to merge all the datasets so that each record represents one purchased product within one order, enriched with:

- Customer information
- Order information
- Product information
- Seller information
- Payment information
- Review information
- Geographic information

#### This integrated dataset will become the foundation for:

- Customer Segmentation
- Sales Forecasting
- Delivery Delay Prediction
- Customer Satisfaction Prediction
- Recommendation System
- Anomaly Detection
- Explainable AI
- Dashboard

### Step 1: Decide the Base Table
- for thisproject the better choice is:Order Items

### Step 2: Integration Strategy
- merge the tables one at a time.
                    ORDER ITEMS
                         │
          ┌──────────────┼──────────────┐
          ▼              ▼              ▼
      ORDERS         PRODUCTS       SELLERS
          │              │
          ▼              ▼
     CUSTOMERS    CATEGORY TRANSLATION
          │
          ▼
     GEOLOCATION
          │
          ▼
      PAYMENTS
          │
          ▼
       REVIEWS

In [60]:
order_items.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14


### Step 3: Merge Orders

In [58]:
retail = order_items.merge(
    orders,
    on="order_id",
    how="left"
)

print(retail.shape)

(112650, 15)


In [59]:
retail.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,delivery_days
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29,3ce436f183e68e07877b285a838db11a,delivered,2017-09-13 08:59:02,2017-09-13 09:45:35,2017-09-19 18:34:16,2017-09-20 23:43:48,2017-09-29,7.0
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93,f6dd3ec061db4e3987629fe6b26e5cce,delivered,2017-04-26 10:53:06,2017-04-26 11:05:13,2017-05-04 14:35:00,2017-05-12 16:04:24,2017-05-15,16.0
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87,6489ae5e4333f3693df5ad4372dab6d3,delivered,2018-01-14 14:33:31,2018-01-14 14:48:30,2018-01-16 12:36:48,2018-01-22 13:19:16,2018-02-05,7.0
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79,d4eb9395c8c0431ee92fce09860c5a06,delivered,2018-08-08 10:00:35,2018-08-08 10:10:18,2018-08-10 13:28:00,2018-08-14 13:32:39,2018-08-20,6.0
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14,58dbd0b2d70206bf40e62cd34e84d795,delivered,2017-02-04 13:57:51,2017-02-04 14:10:13,2017-02-16 09:46:09,2017-03-01 16:42:31,2017-03-17,25.0


### Now every purchased product has:

-purchase date
- delivery date
- status
- customer ID

### Step 4: Merge Customers

In [61]:
retail = retail.merge(
    customers,
    on="customer_id",
    how="left"
)

print(retail.shape)

(112650, 19)


### Now we know

- customer city
- customer state
- customer ZIP prefix

In [62]:
retail.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,delivery_days,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29,3ce436f183e68e07877b285a838db11a,delivered,2017-09-13 08:59:02,2017-09-13 09:45:35,2017-09-19 18:34:16,2017-09-20 23:43:48,2017-09-29,7.0,871766c5855e863f6eccc05f988b23cb,28013,campos dos goytacazes,RJ
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93,f6dd3ec061db4e3987629fe6b26e5cce,delivered,2017-04-26 10:53:06,2017-04-26 11:05:13,2017-05-04 14:35:00,2017-05-12 16:04:24,2017-05-15,16.0,eb28e67c4c0b83846050ddfb8a35d051,15775,santa fe do sul,SP
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87,6489ae5e4333f3693df5ad4372dab6d3,delivered,2018-01-14 14:33:31,2018-01-14 14:48:30,2018-01-16 12:36:48,2018-01-22 13:19:16,2018-02-05,7.0,3818d81c6709e39d06b2738a8d3a2474,35661,para de minas,MG
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79,d4eb9395c8c0431ee92fce09860c5a06,delivered,2018-08-08 10:00:35,2018-08-08 10:10:18,2018-08-10 13:28:00,2018-08-14 13:32:39,2018-08-20,6.0,af861d436cfc08b2c2ddefd0ba074622,12952,atibaia,SP
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14,58dbd0b2d70206bf40e62cd34e84d795,delivered,2017-02-04 13:57:51,2017-02-04 14:10:13,2017-02-16 09:46:09,2017-03-01 16:42:31,2017-03-17,25.0,64b576fb70d441e8f1b2d7d446e483c5,13226,varzea paulista,SP


### Step 5: Merge Products

In [63]:
retail = retail.merge(
    products,
    on="product_id",
    how="left"
)

print(retail.shape)


(112650, 27)


### Now each purchased product has

- category
- weight
- dimensions
- number of photos etc

In [64]:
retail.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,delivery_days,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29,3ce436f183e68e07877b285a838db11a,delivered,2017-09-13 08:59:02,2017-09-13 09:45:35,2017-09-19 18:34:16,2017-09-20 23:43:48,2017-09-29,7.0,871766c5855e863f6eccc05f988b23cb,28013,campos dos goytacazes,RJ,cool_stuff,58.0,598.0,4.0,650.0,28.0,9.0,14.0
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93,f6dd3ec061db4e3987629fe6b26e5cce,delivered,2017-04-26 10:53:06,2017-04-26 11:05:13,2017-05-04 14:35:00,2017-05-12 16:04:24,2017-05-15,16.0,eb28e67c4c0b83846050ddfb8a35d051,15775,santa fe do sul,SP,pet_shop,56.0,239.0,2.0,30000.0,50.0,30.0,40.0
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87,6489ae5e4333f3693df5ad4372dab6d3,delivered,2018-01-14 14:33:31,2018-01-14 14:48:30,2018-01-16 12:36:48,2018-01-22 13:19:16,2018-02-05,7.0,3818d81c6709e39d06b2738a8d3a2474,35661,para de minas,MG,moveis_decoracao,59.0,695.0,2.0,3050.0,33.0,13.0,33.0
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79,d4eb9395c8c0431ee92fce09860c5a06,delivered,2018-08-08 10:00:35,2018-08-08 10:10:18,2018-08-10 13:28:00,2018-08-14 13:32:39,2018-08-20,6.0,af861d436cfc08b2c2ddefd0ba074622,12952,atibaia,SP,perfumaria,42.0,480.0,1.0,200.0,16.0,10.0,15.0
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14,58dbd0b2d70206bf40e62cd34e84d795,delivered,2017-02-04 13:57:51,2017-02-04 14:10:13,2017-02-16 09:46:09,2017-03-01 16:42:31,2017-03-17,25.0,64b576fb70d441e8f1b2d7d446e483c5,13226,varzea paulista,SP,ferramentas_jardim,59.0,409.0,1.0,3750.0,35.0,40.0,30.0


In [65]:
retail.columns

Index(['order_id', 'order_item_id', 'product_id', 'seller_id',
       'shipping_limit_date', 'price', 'freight_value', 'customer_id',
       'order_status', 'order_purchase_timestamp', 'order_approved_at',
       'order_delivered_carrier_date', 'order_delivered_customer_date',
       'order_estimated_delivery_date', 'delivery_days', 'customer_unique_id',
       'customer_zip_code_prefix', 'customer_city', 'customer_state',
       'product_category_name', 'product_name_lenght',
       'product_description_lenght', 'product_photos_qty', 'product_weight_g',
       'product_length_cm', 'product_height_cm', 'product_width_cm'],
      dtype='object')

### Step 6: Translate Product Categories

In [67]:
retail = retail.merge(
    translation,
    on="product_category_name",
    how="left"
)

print(retail.shape)

(112650, 28)


### Now beleza_saude becomes health_beauty

In [68]:
retail.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,delivery_days,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm,product_category_name_english
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29,3ce436f183e68e07877b285a838db11a,delivered,2017-09-13 08:59:02,2017-09-13 09:45:35,2017-09-19 18:34:16,2017-09-20 23:43:48,2017-09-29,7.0,871766c5855e863f6eccc05f988b23cb,28013,campos dos goytacazes,RJ,cool_stuff,58.0,598.0,4.0,650.0,28.0,9.0,14.0,cool_stuff
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93,f6dd3ec061db4e3987629fe6b26e5cce,delivered,2017-04-26 10:53:06,2017-04-26 11:05:13,2017-05-04 14:35:00,2017-05-12 16:04:24,2017-05-15,16.0,eb28e67c4c0b83846050ddfb8a35d051,15775,santa fe do sul,SP,pet_shop,56.0,239.0,2.0,30000.0,50.0,30.0,40.0,pet_shop
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87,6489ae5e4333f3693df5ad4372dab6d3,delivered,2018-01-14 14:33:31,2018-01-14 14:48:30,2018-01-16 12:36:48,2018-01-22 13:19:16,2018-02-05,7.0,3818d81c6709e39d06b2738a8d3a2474,35661,para de minas,MG,moveis_decoracao,59.0,695.0,2.0,3050.0,33.0,13.0,33.0,furniture_decor
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79,d4eb9395c8c0431ee92fce09860c5a06,delivered,2018-08-08 10:00:35,2018-08-08 10:10:18,2018-08-10 13:28:00,2018-08-14 13:32:39,2018-08-20,6.0,af861d436cfc08b2c2ddefd0ba074622,12952,atibaia,SP,perfumaria,42.0,480.0,1.0,200.0,16.0,10.0,15.0,perfumery
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14,58dbd0b2d70206bf40e62cd34e84d795,delivered,2017-02-04 13:57:51,2017-02-04 14:10:13,2017-02-16 09:46:09,2017-03-01 16:42:31,2017-03-17,25.0,64b576fb70d441e8f1b2d7d446e483c5,13226,varzea paulista,SP,ferramentas_jardim,59.0,409.0,1.0,3750.0,35.0,40.0,30.0,garden_tools


### Step 7: Merge Sellers

In [69]:
retail = retail.merge(
    sellers,
    on="seller_id",
    how="left"
)

print(retail.shape)

(112650, 31)


### Now we know

- seller city
- seller state

### Step 8: Merge Payments

#### Notice:

- Payments is one-to-many.

- First aggregate it.


In [70]:
payments_summary = (
    payments
    .groupby("order_id")
    .agg({

        "payment_value":"sum",

        "payment_installments":"max",

        "payment_type":lambda x:x.mode()[0]

    })
    .reset_index()
)

### Merge

In [71]:
retail = retail.merge(

    payments_summary,

    on="order_id",

    how="left"

)

### Why aggregate?

- Some customers paid:

- Credit Card+Voucher+Gift Card

** We ant one payment record per order.

### Step 9: Merge Reviews

* Again, one review per order.

In [73]:
reviews_summary = reviews[
    [

        "order_id",

        "review_score",

        "review_comment_title",

        "review_comment_message"

    ]
]


In [ ]:
# Merge

retail = retail.merge(

    reviews_summary,

    on="order_id",

    how="left"

)

In [77]:
retail.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,delivery_days,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm,product_category_name_english,seller_zip_code_prefix,seller_city,seller_state,payment_value,payment_installments,payment_type,review_score,review_comment_title,review_comment_message
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29,3ce436f183e68e07877b285a838db11a,delivered,2017-09-13 08:59:02,2017-09-13 09:45:35,2017-09-19 18:34:16,2017-09-20 23:43:48,2017-09-29,7.0,871766c5855e863f6eccc05f988b23cb,28013,campos dos goytacazes,RJ,cool_stuff,58.0,598.0,4.0,650.0,28.0,9.0,14.0,cool_stuff,27277,volta redonda,SP,72.19,2.0,credit_card,5.0,,"Perfeito, produto entregue antes do combinado."
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93,f6dd3ec061db4e3987629fe6b26e5cce,delivered,2017-04-26 10:53:06,2017-04-26 11:05:13,2017-05-04 14:35:00,2017-05-12 16:04:24,2017-05-15,16.0,eb28e67c4c0b83846050ddfb8a35d051,15775,santa fe do sul,SP,pet_shop,56.0,239.0,2.0,30000.0,50.0,30.0,40.0,pet_shop,3471,sao paulo,SP,259.83,3.0,credit_card,4.0,,
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87,6489ae5e4333f3693df5ad4372dab6d3,delivered,2018-01-14 14:33:31,2018-01-14 14:48:30,2018-01-16 12:36:48,2018-01-22 13:19:16,2018-02-05,7.0,3818d81c6709e39d06b2738a8d3a2474,35661,para de minas,MG,moveis_decoracao,59.0,695.0,2.0,3050.0,33.0,13.0,33.0,furniture_decor,37564,borda da mata,MG,216.87,5.0,credit_card,5.0,,Chegou antes do prazo previsto e o produto sur...
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79,d4eb9395c8c0431ee92fce09860c5a06,delivered,2018-08-08 10:00:35,2018-08-08 10:10:18,2018-08-10 13:28:00,2018-08-14 13:32:39,2018-08-20,6.0,af861d436cfc08b2c2ddefd0ba074622,12952,atibaia,SP,perfumaria,42.0,480.0,1.0,200.0,16.0,10.0,15.0,perfumery,14403,franca,SP,25.78,2.0,credit_card,4.0,,
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14,58dbd0b2d70206bf40e62cd34e84d795,delivered,2017-02-04 13:57:51,2017-02-04 14:10:13,2017-02-16 09:46:09,2017-03-01 16:42:31,2017-03-17,25.0,64b576fb70d441e8f1b2d7d446e483c5,13226,varzea paulista,SP,ferramentas_jardim,59.0,409.0,1.0,3750.0,35.0,40.0,30.0,garden_tools,87900,loanda,PR,218.04,3.0,credit_card,5.0,,Gostei pois veio no prazo determinado .


### Step 10: Merge Customer Geolocation

In [78]:
retail = retail.merge(

    geo_clean,

    left_on="customer_zip_code_prefix",

    right_on="geolocation_zip_code_prefix",

    how="left"

)

In [80]:
retail.shape

(113314, 42)

### Rename

In [81]:
retail.rename(columns={

    "geolocation_lat":"customer_lat",

    "geolocation_lng":"customer_lng",

    "geolocation_city":"customer_geo_city",

    "geolocation_state":"customer_geo_state"

},inplace=True)

### Step 11: Merge Seller Geolocation

In [82]:
#Step 11: Merge Seller Geolocation
# Create another copy.

seller_geo = geo_clean.rename(columns={

    "geolocation_zip_code_prefix":"seller_zip_code_prefix",

    "geolocation_lat":"seller_lat",

    "geolocation_lng":"seller_lng",

    "geolocation_city":"seller_geo_city",

    "geolocation_state":"seller_geo_state"

})


In [83]:
# Merge

retail = retail.merge(

    seller_geo,

    on="seller_zip_code_prefix",

    how="left"

)
retail.shape

(113314, 46)

### Step 12: Check the Shape

In [84]:
print(retail.shape)

(113314, 46)


### Step 13: Inspect the Dataset

In [85]:
retail.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,delivery_days,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm,product_category_name_english,seller_zip_code_prefix,seller_city,seller_state,payment_value,payment_installments,payment_type,review_score,review_comment_title,review_comment_message,geolocation_zip_code_prefix,customer_lat,customer_lng,customer_geo_city,customer_geo_state,seller_lat,seller_lng,seller_geo_city,seller_geo_state
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29,3ce436f183e68e07877b285a838db11a,delivered,2017-09-13 08:59:02,2017-09-13 09:45:35,2017-09-19 18:34:16,2017-09-20 23:43:48,2017-09-29,7.0,871766c5855e863f6eccc05f988b23cb,28013,campos dos goytacazes,RJ,cool_stuff,58.0,598.0,4.0,650.0,28.0,9.0,14.0,cool_stuff,27277,volta redonda,SP,72.19,2.0,credit_card,5.0,,"Perfeito, produto entregue antes do combinado.",28013.0,-21.763186,-41.310265,campos dos goytacazes,RJ,-22.497188,-44.127324,volta redonda,RJ
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93,f6dd3ec061db4e3987629fe6b26e5cce,delivered,2017-04-26 10:53:06,2017-04-26 11:05:13,2017-05-04 14:35:00,2017-05-12 16:04:24,2017-05-15,16.0,eb28e67c4c0b83846050ddfb8a35d051,15775,santa fe do sul,SP,pet_shop,56.0,239.0,2.0,30000.0,50.0,30.0,40.0,pet_shop,3471,sao paulo,SP,259.83,3.0,credit_card,4.0,,,15775.0,-20.222506,-50.898951,santa fe do sul,SP,-23.565754,-46.519097,sao paulo,SP
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87,6489ae5e4333f3693df5ad4372dab6d3,delivered,2018-01-14 14:33:31,2018-01-14 14:48:30,2018-01-16 12:36:48,2018-01-22 13:19:16,2018-02-05,7.0,3818d81c6709e39d06b2738a8d3a2474,35661,para de minas,MG,moveis_decoracao,59.0,695.0,2.0,3050.0,33.0,13.0,33.0,furniture_decor,37564,borda da mata,MG,216.87,5.0,credit_card,5.0,,Chegou antes do prazo previsto e o produto sur...,35661.0,-19.869998,-44.593059,para de minas,MG,-22.262802,-46.170735,borda da mata,MG
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79,d4eb9395c8c0431ee92fce09860c5a06,delivered,2018-08-08 10:00:35,2018-08-08 10:10:18,2018-08-10 13:28:00,2018-08-14 13:32:39,2018-08-20,6.0,af861d436cfc08b2c2ddefd0ba074622,12952,atibaia,SP,perfumaria,42.0,480.0,1.0,200.0,16.0,10.0,15.0,perfumery,14403,franca,SP,25.78,2.0,credit_card,4.0,,,12952.0,-23.105968,-46.590277,atibaia,SP,-20.553651,-47.387145,franca,SP
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14,58dbd0b2d70206bf40e62cd34e84d795,delivered,2017-02-04 13:57:51,2017-02-04 14:10:13,2017-02-16 09:46:09,2017-03-01 16:42:31,2017-03-17,25.0,64b576fb70d441e8f1b2d7d446e483c5,13226,varzea paulista,SP,ferramentas_jardim,59.0,409.0,1.0,3750.0,35.0,40.0,30.0,garden_tools,87900,loanda,PR,218.04,3.0,credit_card,5.0,,Gostei pois veio no prazo determinado .,13226.0,-23.243402,-46.827614,varzea paulista,SP,-22.929583,-53.135750,loanda,PR


In [86]:
retail.columns

Index(['order_id', 'order_item_id', 'product_id', 'seller_id',
       'shipping_limit_date', 'price', 'freight_value', 'customer_id',
       'order_status', 'order_purchase_timestamp', 'order_approved_at',
       'order_delivered_carrier_date', 'order_delivered_customer_date',
       'order_estimated_delivery_date', 'delivery_days', 'customer_unique_id',
       'customer_zip_code_prefix', 'customer_city', 'customer_state',
       'product_category_name', 'product_name_lenght',
       'product_description_lenght', 'product_photos_qty', 'product_weight_g',
       'product_length_cm', 'product_height_cm', 'product_width_cm',
       'product_category_name_english', 'seller_zip_code_prefix',
       'seller_city', 'seller_state', 'payment_value', 'payment_installments',
       'payment_type', 'review_score', 'review_comment_title',
       'review_comment_message', 'geolocation_zip_code_prefix', 'customer_lat',
       'customer_lng', 'customer_geo_city', 'customer_geo_state', 'seller_lat',
  

### Step 14: Check Missing Values Again

In [88]:
missing = (
    retail.isnull()
    .sum()
    .sort_values(ascending=False)
)

missing[missing > 0]

order_delivered_customer_date    2475
delivery_days                    2475
product_category_name_english    1636
order_delivered_carrier_date     1203
review_comment_message            942
review_score                      942
review_comment_title              942
customer_geo_state                307
customer_geo_city                 307
customer_lng                      307
customer_lat                      307
geolocation_zip_code_prefix       307
seller_geo_state                  257
seller_lat                        257
seller_lng                        257
seller_geo_city                   257
order_approved_at                  15
payment_type                        3
payment_installments                3
payment_value                       3
dtype: int64

| Column                          | Action                                                      | Why                             |
| ------------------------------- | ----------------------------------------------------------- | ------------------------------- |
| `order_delivered_customer_date` | Keep NaN                                                    | Order not delivered             |
| `delivery_days`                 | Keep NaN                                                    | Cannot compute                  |
| `order_delivered_carrier_date`  | Keep NaN                                                    | Not shipped                     |
| `order_approved_at`             | Keep NaN                                                    | Rare anomalies/business process |
| `product_category_name_english` | Fill `"Unknown"` (if corresponding category is `"Unknown"`) | Missing translation             |
| `review_comment_title`          | Fill `""`                                                   | Optional text                   |
| `review_comment_message`        | Fill `""`                                                   | Optional text                   |
| `review_score`                  | Keep NaN                                                    | No review submitted             |
| Customer geolocation            | Keep NaN                                                    | ZIP prefix not found            |
| Seller geolocation              | Keep NaN                                                    | ZIP prefix not found            |
| `payment_type`                  | Fill `"Unknown"` if truly no payment                        | Business meaning                |
| `payment_value`                 | Fill `0` only if no payment exists                          | Business meaning                |
| `payment_installments`          | Fill `0` only if no payment exists                          | Business meaning                |


### Step 15: Save the Master Dataset

In [89]:
retail.to_csv(
    "Retail_Master_Dataset.csv",
    index=False
)

## Preprocessing - Missing Values

### Product Category Translation
- product_category_name_english Missing = 1,636
- This does not necessarily mean the product category is missing.

** Check first:

In [90]:
retail[
    retail["product_category_name_english"].isnull()
][["product_category_name"]].drop_duplicates()

- There are two possibilities:
*Case 1: product_category_name = Unknown
- Then the translation table simply has no translation for "Unknown".
- Fill it:

In [ ]:
retail["product_category_name_english"] = (
    retail["product_category_name_english"]
    .fillna("Unknown")
)

** Case 2

- The Portuguese category exists but has no matching translation.

- Then keep the Portuguese value or map it manually

In [91]:
retail.loc[
    retail["product_category_name_english"].isnull(),
    "product_category_name"
].value_counts(dropna=False)

Series([], Name: product_category_name, dtype: int64)

In [92]:
retail["product_category_name_english"].isnull().sum()

0

In [93]:
retail[
    [
        "product_category_name",
        "product_category_name_english"
    ]
].sample(10)

,product_category_name,product_category_name_english
113037,telefonia,telephony
42394,beleza_saude,health_beauty
96045,cama_mesa_banho,bed_bath_table
20271,cool_stuff,cool_stuff
51140,moveis_decoracao,furniture_decor
109036,informatica_acessorios,computers_accessories
90065,Unknown,Unknown
4379,informatica_acessorios,computers_accessories
45843,livros_interesse_geral,books_general_interest
79378,telefonia,telephony


In [94]:
retail["product_category_name"].isnull().sum()

0

### Delivery Features
*Suppose delivery_days contains NaN.
- Don't replace it.
- Only calculate delivery features for completed orders.


In [95]:
delivery_df = retail[
    retail["order_status"] == "delivered"
].copy()

### Review

- Add a review indicator to create a flag indicating whether a customer left a review.

In [102]:
retail["has_review"] = retail["review_score"].notna().astype(int)

### Payment Information
*Only 3 missing values?

In [96]:
# Investigate first

retail[
    retail["payment_value"].isnull()
][
[
"order_status",
"order_id"
]
]

,order_status,order_id
84898,delivered,bfbd0f9bdef84302105ad712db648a6c
84899,delivered,bfbd0f9bdef84302105ad712db648a6c
84900,delivered,bfbd0f9bdef84302105ad712db648a6c


In [97]:
# Check whether the payment exists

payments[
    payments["order_id"] == "bfbd0f9bdef84302105ad712db648a6c"
]

,order_id,payment_sequential,payment_type,payment_installments,payment_value


### Check the aggregated payments


In [98]:
payments_summary[
    payments_summary["order_id"] == "bfbd0f9bdef84302105ad712db648a6c"
]

,order_id,payment_value,payment_installments,payment_type


In [99]:
payments[
    payments["order_id"] == "bfbd0f9bdef84302105ad712db648a6c"
]

,order_id,payment_sequential,payment_type,payment_installments,payment_value


In [100]:
payments["order_id"].eq("bfbd0f9bdef84302105ad712db648a6c").sum()

0

# Business insight
#### During data integration, one delivered order (bfbd0f9bdef84302105ad712db648a6c) was found to have no corresponding payment record in the original payments dataset. As payment information was required for downstream analyses, this anomalous order (three associated order-item records) was removed from the master dataset. The removal affected less than 0.003% of the data and therefore had no material impact on the overall analysis.

In [101]:
# Removing the anomalous data
retail = retail[
    retail["order_id"] != "bfbd0f9bdef84302105ad712db648a6c"
].copy()

### Check Missing Values Again

In [103]:
missing = (
    retail.isnull()
    .sum()
    .sort_values(ascending=False)
)

missing[missing > 0]

order_delivered_customer_date    2475
delivery_days                    2475
order_delivered_carrier_date     1203
review_comment_message            942
review_score                      942
review_comment_title              942
customer_geo_state                307
customer_geo_city                 307
customer_lng                      307
customer_lat                      307
geolocation_zip_code_prefix       307
seller_geo_state                  257
seller_geo_city                   257
seller_lng                        257
seller_lat                        257
order_approved_at                  15
dtype: int64

| Column                          | Missing | % of ~112,000 rows | Action                          |
| ------------------------------- | ------: | -----------------: | ------------------------------- |
| `order_delivered_customer_date` |   2,475 |              ~2.2% | ✅ Keep                          |
| `delivery_days`                 |   2,475 |              ~2.2% | ✅ Keep                          |
| `order_delivered_carrier_date`  |   1,203 |              ~1.1% | ✅ Keep                          |
| `review_score`                  |     942 |              ~0.8% | ✅ Keep                          |
| `review_comment_title`          |     942 |              ~0.8% | ⚠️ Check if they are really NaN |
| `review_comment_message`        |     942 |              ~0.8% | ⚠️ Check if they are really NaN |
| Customer geolocation            |     307 |              ~0.3% | ✅ Keep                          |
| Seller geolocation              |     257 |              ~0.2% | ✅ Keep                          |
| `order_approved_at`             |      15 |             ~0.01% | ✅ Keep                          |


review_comment_title- Check if they are really NaN
review_comment_message- Check if they are really NaN

### 1. Review comments

In [104]:
retail["review_comment_title"] = (
    retail["review_comment_title"].fillna("")
)

retail["review_comment_message"] = (
    retail["review_comment_message"].fillna("")
)

In [105]:
retail["review_comment_title"].isnull().sum()

0

In [106]:
retail["review_comment_message"].isnull().sum()

0

### 2. Review score

- Leave it.

- NaN = customer never submitted a review

- This is information, not an error.

### 3. Delivery columns

Leave them.

They correspond to

cancelled
unavailable
processing
shipped

orders.

Do not fill them.

### 4. Customer geolocation

Only 307 rows.

Leave them.

I cannot invent GPS coordinates.
### 5. Seller geolocation

Same.

Leave them.

### 6. order_approved_at

Only 15 rows.

This is negligible.

Leave them.

### Check Missing Values Again

In [108]:
missing = (
    retail.isnull()
    .sum()
    .sort_values(ascending=False)
)

missing[missing > 0]

order_delivered_customer_date    2475
delivery_days                    2475
order_delivered_carrier_date     1203
review_score                      942
customer_geo_city                 307
geolocation_zip_code_prefix       307
customer_lng                      307
customer_lat                      307
customer_geo_state                307
seller_lat                        257
seller_lng                        257
seller_geo_city                   257
seller_geo_state                  257
order_approved_at                  15
dtype: int64

### At this point, the Data Cleaning and Data Integration phases are complete. The remaining missing values are all explainable from the business process and should be documented rather than imputed.

### The next major phase should be Feature Engineering, where I'll derive business features such as RFM metrics, customer lifetime value, delivery performance, seller performance, product popularity, and payment behavior. These engineered features will provide much more value to the machine learning models than further attempts to eliminate the remaining missing values.

### Save a new version of the dataset which is good for version control

In [109]:
retail.to_csv("Retail_Master_Dataset_Cleaned.csv", index=False)

print("Retail_Master_Dataset_Cleaned.csv saved successfully!")

Retail_Master_Dataset_Cleaned.csv saved successfully!


### Verify that it was saved correctly

In [111]:
import os

print(os.getcwd())  # Shows the current working directory


C:\Users\tiwal\Desktop\OSAP\CIMT\CAPSTONE II


In [112]:
#Then check that the file exists:

os.path.exists("Retail_Master_Dataset.csv")

True

In [ ]:
EDA — what to visualise and analyse:
•	Order volume by month and year — identify seasonality and growth trends.
•	Revenue distribution by product category — find the top-performing categories.
•	Delivery performance: calculate actual vs. estimated delivery time for every order. What percentage are late?
•	Review score distribution — is the dataset skewed toward positive reviews?
•	Geographic distribution of customers and sellers using geolocation data.
•	Payment method breakdown: credit card vs. boleto vs. voucher vs. debit card.
•	Correlation heatmap of numerical features.

Deliverable: 01_EDA.ipynb with a clean master DataFrame, summary statistics, and at least 8 visualisations that tell a clear business story about NovaMart's data.
